brave/src/main/java/com/linecorp/decaton/processor/runtime/BraveTracingProvider.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.runtime;
+
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import brave.Span;
+import brave.kafka.clients.KafkaTracing;
+import brave.propagation.CurrentTraceContext.Scope;
+
+public class BraveTracingProvider implements TracingProvider<Span> {
+    private static final ThreadLocal<Scope> CURRENT_SCOPE = new ThreadLocal<>();
+    private final KafkaTracing kafkaTracing;
+
+    public BraveTracingProvider(KafkaTracing kafkaTracing) {
+        this.kafkaTracing = kafkaTracing;
+    }
+
+    @Override
+    public Span traceFor(ConsumerRecord<?, ?> record) {
+        return kafkaTracing.nextSpan(record).name("decaton.process").start();
+    }
+
+    @Override
+    public void processingStart(Span trace) {
+        CURRENT_SCOPE.set(
+                kafkaTracing.messagingTracing().tracing().currentTraceContext().newScope(trace.context()));
+        trace.annotate("decaton.pipeline.start");
```

Comment:
I'm not enough experienced with zipkin/brave so not attempting to doubt something but just to learn: what's the criteria of deciding whether to use span and annotation?

Reply:
I'm not an expert either, but I think the big difference is that a span has its own tags, whereas an annotation is basically just a timestamp on an existing span. So e.g. if we wanted to set some kind of processor id as a tag then we'd need to make a span for each processor.

Reply:
Thanks for explanation :)

> if we wanted to set some kind of processor id as a tag then we'd need to make a span for each processor.

and this made me to came up with another question; shouldn't we pass 'subscriptionId' to `traceFor` method so that the driver can set it as a tag in span? otherwise how can one distinguish which decaton instance does the trace come from?

Reply:
Good point - will add that.